### PR TITLE
Music DAO MVP

### PR DESCRIPTION
**Pull request by brian2509**
(draft, should not be merged yet)

New version of MusicDAO including a number of version bumps for kotlin and dependencies. 

Most important dependency updates:
- kotlin 1.4.30 > 1.5.21
- gradle: 6.5 > 6.8
- compileSdkVersion: 30 > 31
- com.android.tools.build:gradle > com.android.tools.build:gradle: 4.1.3 > 4.2.0
- nav_version: 2.2.1 > 2.3.1

Changes outside of musicdao module should be limited to:
- suppressing deprecation warnings `@Suppress("DEPRECATION")` in multiple modules 
- fixing (simple) tests which were failing
- change usage of (now) private animations in xml files throughout project due to  [all "@anim/*" being private now], replaced with `@animator` fade animations (https://stackoverflow.com/questions/67305830/where-is-anim-fragment-fade-enter-included-in)

### CODE REVIEW

**Comment by synctext** [comment]
118 commits!
Great additional tech for upcoming Master course🦄🦄🦄

File: app/build.gradle [hunk 1]
```diff
@@ -103,7 +109,9 @@ allprojects {
     repositories {
         mavenCentral()
         jcenter()
-
```
Comment (InvictusRMC) [inline app/build.gradle:106]: This shouldn't be necessary

File: app/build.gradle [hunk 2]
```diff
@@ -6,15 +6,21 @@ apply plugin: 'org.jlleitschuh.gradle.ktlint'
 apply plugin: 'com.google.gms.google-services'
 apply plugin: 'com.google.firebase.crashlytics'
 apply plugin: 'kotlinx-serialization'
+apply plugin: "dagger.hilt.android.plugin"
+apply plugin: "kotlin-kapt"
 
 buildscript {
+    ext.hilt_ver = '2.44.2'
```
Comment (InvictusRMC) [inline app/build.gradle:13]: Define this in global build.gradle

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -146,17 +146,12 @@
             android:windowSoftInputMode="stateAlwaysHidden" />
 
         <activity
-            android:name="com.example.musicdao.MusicService"
+            android:name="com.example.musicdao.MusicActivity"
```
Comment (InvictusRMC) [inline app/src/main/AndroidManifest.xml:149]: could you replace this `com.example`? Looks really weird. Shouldn't have been merged like this in the first place.

File: currencyii/src/main/java/nl/tudelft/trustchain/currencyii/util/DAOJoinHelper.kt [hunk 4]
```diff
@@ -157,14 +154,14 @@ class DAOJoinHelper {
 
         if (status) {
             Log.i("Coin", "Successfully submitted taproot transaction to server")
-            activity.runOnUiThread {
-                Toast.makeText(context, "Successfully submitted the transaction", Toast.LENGTH_SHORT).show()
-            }
+//            activity.runOnUiThread {
```
Comment (InvictusRMC) [inline currencyii/src/main/java/nl/tudelft/trustchain/currencyii/util/DAOJoinHelper.kt:157]: remove this?

File: currencyii/src/main/java/nl/tudelft/trustchain/currencyii/util/DAOTransferFundsHelper.kt [hunk 5]
```diff
@@ -132,15 +132,15 @@ class DAOTransferFundsHelper {
...
             activity.runOnUiThread {
                 Toast.makeText(context, "Successfully submitted the transaction", Toast.LENGTH_SHORT).show()
             }
         } else {
-            Log.e("Coin", "taproot transaction submission to server failed")
-            activity.runOnUiThread {
-                Toast.makeText(context, "Failed to submit the transaction to the server", Toast.LENGTH_SHORT).show()
-            }
+            Log.d("MVDAO", "taproot transaction submission to server failed")
+//            activity.runOnUiThread {
```
Comment (InvictusRMC) [inline currencyii/src/main/java/nl/tudelft/trustchain/currencyii/util/DAOTransferFundsHelper.kt:141]: Remove this?

File: debug/src/test/java/nl/tudelft/trustchain/debug/ExampleUnitTest.kt [hunk 6]
```diff
@@ -11,6 +11,6 @@ import org.junit.Test
 class ExampleUnitTest {
```
Comment (InvictusRMC) [inline debug/src/test/java/nl/tudelft/trustchain/debug/ExampleUnitTest.kt:11]: Might as well remove these files.

File: gossipML/build.gradle [hunk 7]
```diff
@@ -80,7 +80,7 @@ android {
         abi {
             enable true
             reset()
-            include 'arm64-v8a', 'x86', 'armeabi-v7a', 'x86_64'
+            include 'x86', 'armeabi-v7a', 'x86_64'
```
Comment (InvictusRMC) [inline gossipML/build.gradle:83]: Why did you remove arm64 architecture?

File: musicdao/build.gradle [hunk 8]
```diff
@@ -126,6 +190,13 @@ tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
     ]
 }
 
+//tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).configureEach {
```
Comment (InvictusRMC) [inline musicdao/build.gradle:193]: Remove if not needed

File: musicdao/src/main/java/com/example/musicdao/HiltModules.kt [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        @ApplicationContext context: Context
+    ): WalletManager {
+        Log.d("MVDAO", "INITIATING DAO MODULE.")
+        if (WalletManagerAndroid.isInitialized()) {
+            Log.d("MVDAO", "DAO MODULE ALREADY INITIALIZED, SKIPPING")
+            return WalletManagerAndroid.getInstance()
+        }
+        val params = BitcoinNetworkOptions.REG_TEST
+
+//        val seed = WalletManager.generateRandomDeterministicSeed(params)
```
Comment (InvictusRMC) [inline musicdao/src/main/java/com/example/musicdao/HiltModules.kt:159]: Remove

File: musicdao/src/main/java/com/example/musicdao/core/repositories/SwarmHealthRepository.kt [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import nl.tudelft.ipv8.android.IPv8Android
+
+class SwarmHealthRepository(
+    private val sessionManager: SessionManager,
+    val musicCommunity: MusicCommunity
+) {
+
+    var mergedSwarmHealth: MutableMap<Sha1Hash, SwarmHealth> = mutableMapOf()
+
+//    val localSwarmHealthMap: MutableMap<Sha1Hash, SwarmHealth>
```
Comment (InvictusRMC) [inline musicdao/src/main/java/com/example/musicdao/core/repositories/SwarmHealthRepository.kt:16]: Remove

File: musicdao/src/main/java/com/example/musicdao/core/repositories/album/CreateReleaseUseCase.kt [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        val releaseId = UUID.randomUUID().toString()
+        Log.d("MusicDao", "CreateReleaseUseCase: $releaseId")
+
+        val root = torrentEngine.simulateDownload(context, uris)
+        if (root == null) {
+            Log.d("MusicDao", "CreateReleaseUseCase: could not simulate download")
+            return false
+        }
+
+//        val contentFolder = TorrentEngine.rootToContentFolder(root) ?: return false
```
Comment (InvictusRMC) [inline musicdao/src/main/java/com/example/musicdao/core/repositories/album/CreateReleaseUseCase.kt:37]: Remove

**Review by InvictusRMC — COMMENTED** [review]
Please remove your new (old) libtorrent binaries and use the newer ones used in master and rename the `com.example`. Overall, code requires some clean up.

File: musicdao/build.gradle [hunk 12]
```diff
@@ -43,19 +61,27 @@ android {
...
     packagingOptions {
         resources {
             excludes += ['META-INF/DEPENDENCIES', 'META-INF/LICENSE', 'META-INF/LICENSE.txt', 'META-INF/license.txt', 'META-INF/NOTICE', 'META-INF/NOTICE.txt', 'META-INF/notice.txt', 'META-INF/ASL2.0', 'META-INF/*.kotlin_module']
         }
     }
 
-    namespace 'com.example.musicdao'
+    composeOptions {
+        kotlinCompilerVersion kotlin_version
+        kotlinCompilerExtensionVersion "1.3.1"
```
Comment (InvictusRMC) [inline musicdao/build.gradle:73]: Isn't this an old version?

File: musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/dao/DaoCreateScreen.kt [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    Card(
+        modifier = Modifier
+            .padding(20.dp)
+    ) {
+        Column(
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(20.dp)
+        ) {
+//            OutlinedTextField(
```
Comment (InvictusRMC) [inline musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/dao/DaoCreateScreen.kt:34]: Remove commented code

File: musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/dao/DaoViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,664 @@
...
+
+    @SuppressLint("NewApi")
+    private suspend fun transferFundsClickedLong(
+        bitcoinPublicKey: String,
+        satoshiTransferAmount: Long,
+        blockHash: ByteArray,
+        context: Context,
+        activityRequired: Activity
+    ) {
+//        if (!validateTransferInput()) {
```
Comment (InvictusRMC) [inline musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/dao/DaoViewModel.kt:407]: Remove commented code

File: musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/dao/DaoViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,664 @@
...
+
+            Log.d(
+                "MVDAO",
+                "Funds transferred!"
+            )
+            SnackbarHandler.displaySnackbar("Funds have been transferred!")
+            refreshOneShot()
+        } catch (t: Throwable) {
+            Log.d("MVDAO", "Transferring funds failed. ${t.message ?: "No further information"}.")
+//            resetWalletInitializationValues()
```
Comment (InvictusRMC) [inline musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/dao/DaoViewModel.kt:487]: remove commented code

File: musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/dao/DaoViewModel.kt [hunk 16]
```diff
@@ -0,0 +1,664 @@
...
+            "MVDAO",
+            "Responses for ${blockData.SW_UNIQUE_ID}.${blockData.SW_UNIQUE_PROPOSAL_ID}: ${responses.size}/$requiredSignatures"
+        )
+
+        Log.d(
+            "MVDAO",
+            "Collecting signatures: ${responses.size}/${blockData.SW_SIGNATURES_REQUIRED} received!"
+        )
+
+//        SnackbarHandler.displaySnackbar("Collecting signatures: ${responses.size}/${blockData.SW_SIGNATURES_REQUIRED} received!")
```
Comment (InvictusRMC) [inline musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/dao/DaoViewModel.kt:516]: Remove commented code

File: musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/dao/ProposalCreateScreen.kt [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+
+    val local = LocalContext.current
+
+    val dao = daoViewModel.getDao(daoId)
+
+    var expanded by remember { mutableStateOf(false) }
+
+    fun newProposal() {
+        if (dao != null) {
+            // validaation
```
Comment (InvictusRMC) [inline musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/dao/ProposalCreateScreen.kt:34]:
```suggestion
            // validation
```

File: musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/home/HomeScreen.kt [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+@RequiresApi(Build.VERSION_CODES.O)
+@ExperimentalMaterialApi
+@OptIn(ExperimentalFoundationApi::class)
+@Composable
+fun HomeScreen(
+    navController: NavHostController,
+    homeScreenViewModel: HomeScreenViewModel,
+    screenViewModel: SearchScreenViewModel
+) {
+//    val releasesState by homeScreenViewModel.releases.observeAsState(listOf())
```
Comment (InvictusRMC) [inline musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/home/HomeScreen.kt:35]: Remove commented code

File: musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/home/HomeScreen.kt [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+                EmptyState(
+                    firstLine = "No releases found",
+                    secondLine = "Make a release yourself or wait for releases to come in"
+                )
+            }
+        }
+
+        val modifier = if (releases.isEmpty()) Modifier else Modifier.fillMaxSize()
+
+//        Column(modifier = Modifier.fillMaxSize()) {
```
Comment (InvictusRMC) [inline musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/home/HomeScreen.kt:81]: Remove commented code

**Review by InvictusRMC — CHANGES_REQUESTED** [review]
Your app does not build on Mac M1, this seems to have something to do with Room: https://issuetracker.google.com/issues/174695268

```
Caused by: java.lang.Exception: No native library is found for os.name=Mac and os.arch=aarch64. path=/org/sqlite/native/Mac/aarch64
	at org.sqlite.SQLiteJDBCLoader.loadSQLiteNativeLibrary(SQLiteJDBCLoader.java:333)
	at org.sqlite.SQLiteJDBCLoader.initialize(SQLiteJDBCLoader.java:64)
	at androidx.room.verifier.DatabaseVerifier.<clinit>(DatabaseVerifier.kt:71)
	... 53 more
```

**Comment by synctext** [comment]
@brian2509 please focus and finish this PR. As was requested back in Jan, we need this code to work for all Blockchain Engineering master students. Next week this course already starts!
So the deadline of getting it on Google Play store has already passed :fearful:

**Comment by InvictusRMC** [comment]
I fixed the dependency issues and verified that the APK builds. If you fix my requested changes, this can be merged!

File: musicdao/build.gradle (hunk 12 above)
Comment (brian2509) [reply musicdao/build.gradle:73]: Latest compatible version with Kotlin 1.7.10 (which we are using) is 1.3.1.
https://developer.android.com/jetpack/androidx/releases/compose-compiler

```
e: java.lang.Error: Something went wrong while checking for version compatibility between the Compose Compiler and the Kotlin Compiler.  It is possible that the versions are incompatible.  Please verify your kotlin version  and consult the Compose-Kotlin compatibility map located at https://developer.android.com/jetpack/androidx/releases/compose-kotlin
```

File: app/build.gradle [hunk 20]
```diff
@@ -180,6 +183,18 @@ dependencies {
     testImplementation 'junit:junit:4.13.2'
     androidTestImplementation 'androidx.test.ext:junit:1.1.5'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'
+
+    // Hilt
```
[the remark this answers is not in the document]
Comment (brian2509) [reply app/build.gradle:187]: Unfortunately it is due to the fact I want to use Android specific Hilt features, which requires me to[ annotate the Application class with @HiltAndroidApp](https://developer.android.com/training/dependency-injection/hilt-android), thus the main application should have Hilt included.

File: currencyii/src/main/java/nl/tudelft/trustchain/currencyii/util/DAOJoinHelper.kt [hunk 21]
```diff
@@ -93,7 +91,7 @@ class DAOJoinHelper {
      * Create a bitcoin transaction that creates a new shared wallet. This takes some time to complete.
      *
      * NOTE:
-     *  - the latest walletBlockData should be given, otherwise the serialized transaction is invalid.
+     *  - the latest walletBlockData should be  given, otherwise the serialized transaction is invalid.
```
Comment (InvictusRMC) [inline currencyii/src/main/java/nl/tudelft/trustchain/currencyii/util/DAOJoinHelper.kt:94]:
```suggestion
     *  - the latest walletBlockData should be given, otherwise the serialized transaction is invalid.
```

File: currencyii/src/main/java/nl/tudelft/trustchain/currencyii/util/DAOJoinHelper.kt [hunk 22]
```diff
@@ -36,7 +34,7 @@ class DAOJoinHelper {
      * 2.1 Send a proposal on the trust chain to join a shared wallet and to collect signatures.
      * The proposal is a serialized bitcoin join transaction.
      * **NOTE** the latest walletBlockData should be given, otherwise the serialized transaction is invalid.
-     * @param mostRecentWalletBlock - the latest (that you know of) shared wallet block.
+     * @param mostRecentWalletBlock - the latest (that you know of) sharedwallet block.
```
Comment (InvictusRMC) [inline currencyii/src/main/java/nl/tudelft/trustchain/currencyii/util/DAOJoinHelper.kt:37]:
```suggestion
     * @param mostRecentWalletBlock - the latest (that you know of) shared wallet block.
```

File: musicdao/src/main/java/nl/tudelft/trustchain/musicdao/core/ipv8/blocks/release_publish/ReleasePublishBlockValidator.kt [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+    fun validateTransaction(transaction: TrustChainTransaction): Boolean {
+        val releaseId = transaction["releaseId"]
+        val magnet = transaction["magnet"]
+        val title = transaction["title"]
+        val artist = transaction["artist"]
+        val publisher = transaction["publisher"]
+        val releaseDate = transaction["releaseDate"]
+        val protocolVersion = transaction["protocolVersion"]
+
+//        Log.d("MusicDao", "Validating: $protocolVersion === ${Constants.PROTOCOL_VERSION} = ${protocolVersion == Constants.PROTOCOL_VERSION}")
```
Comment (InvictusRMC) [inline musicdao/src/main/java/nl/tudelft/trustchain/musicdao/core/ipv8/blocks/release_publish/ReleasePublishBlockValidator.kt:37]:
```suggestion
```

File: musicdao/src/main/java/nl/tudelft/trustchain/musicdao/core/ipv8/blocks/release_publish/ReleasePublishBlockValidator.kt [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+
+    fun validateTransaction(transaction: TrustChainTransaction): Boolean {
+        val releaseId = transaction["releaseId"]
+        val magnet = transaction["magnet"]
+        val title = transaction["title"]
+        val artist = transaction["artist"]
+        val publisher = transaction["publisher"]
+        val releaseDate = transaction["releaseDate"]
+        val protocolVersion = transaction["protocolVersion"]
+
```
Comment (InvictusRMC) [inline musicdao/src/main/java/nl/tudelft/trustchain/musicdao/core/ipv8/blocks/release_publish/ReleasePublishBlockValidator.kt:36]:
```suggestion
```

File: musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/release/ReleaseScreen.kt [hunk 25]
```diff
@@ -0,0 +1,344 @@
...
+                }
+            }
+            IconButton(onClick = { /*TODO*/ }) {
+                Icon(
+                    imageVector = Icons.Outlined.PlayArrow,
+                    contentDescription = null,
+
+                )
+            }
+//            OutlinedButton(onClick = { navController.navigate(Screen.Donate.route) }) {
```
Comment (InvictusRMC) [inline musicdao/src/main/java/nl/tudelft/trustchain/musicdao/ui/screens/release/ReleaseScreen.kt:337]: Remove this

**Review by InvictusRMC — CHANGES_REQUESTED** [review]
Remove `musicdao-datafeeder/libs/jlibtorrent-1.2.15.2.jar`

Minor final requested changes!

**Review by InvictusRMC — APPROVED** [review]
🎉
